### PR TITLE
packaging: add IBM power support

### PR DESCRIPTION
**Pull request by sumitd2**
IBM power 64 bit little endian support for fluent-bit

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [NA] Example configuration file for the change
- [Y] Debug log output from testing the change

- [Y] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [NA] Attached [local packaging test](./packaging/local-build-all.sh) output showing all targets (including any new ones) build.

**Documentation**

- [NA] Documentation required for this feature

NA

**Backporting**

- [NA] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by sumitd2** [comment]
[fluent-bit.log](https://github.com/fluent/fluent-bit/files/9416214/fluent-bit.log)

**Comment by sumitd2** [comment]
Dear team,

There are four failing tests on ppc64le:
         20 - flb-rt-filter_stdout (Failed)
         57 - flb-rt-out_td (Failed)
         69 - flb-it-network (Failed)
        106 - flb-it-lua (Failed)
Out of these, the first three are also failing on Intel, which we believe are only minor issues. Once you confirm this PR can, in principle, be merged we could look at the failing flb-it-lua test.

**Comment by edsiper** [comment]
hi @sumitd2 

first of all thanks for this contribution!

I have a couple of questions:

- are you part of IBM ?, if so, is your intention to keep maintaining the support with the architecture ?

- where that luajit2/ comes from ? can you point me to the original source code project

- filter_stdout should not fail since it just prints to stdout, no extra operation involved.

happy to join a call if that simplifies everything

**Comment by leonardo-albertovich** [comment]
From what I could see the changes to fluent-bit were pretty minimal (something in utf-8 processing and libco), wouldn't it be a viable approach to disable the JIT functionality in PPC to keep the footprint small and then figure out if we want to have full support based on demand?

**Comment by sumitd2** [comment]
Hi @edsiper,

Thank you for your quick response. 

* Yes we are from the IBM Power porting team. We mostly just add Power support to repos, but I will have to check if there  any kind of support we can provide. 
You will have to first enable Power in your CI. Github Actions doesnt support Power, but we have enabled GitHub actions for other repositories using qemu / buildx. Will have to evaluate if that can be done here. Also, Travis CI supports power.

* You can find luajit2 here: https://github.com/openresty/luajit2

* filter_stdout is failing on Intel as well, it seems broken. It was passing on 1.9.5 last time we checked.

**Comment by sumitd2** [comment]
Hi @leonardo-albertovich , let me know if you guys want JIT disabled for ppc64le.

**Comment by leonardo-albertovich** [comment]
It's up to @edsiper, I personally think that the footprint of the PR without the luajit2 addition would be much more manageable, especially since I think that at the moment we are not using the actual JIT mode in luajit but I could be wrong about that since I'm not familiar enough with the component.

File: lib/luajit2/Makefile
Comment (cosmo0920) [inline lib/luajit2/Makefile:21]: This is just my curious. Can we use the recent commit of LuaJIT for ppc64le platform?

I don't have real ppc64le environment but it seems that this luajit2 libraries should be luajit 2.1.0-beta3 version.
This seems to be a bit of old than current fbit using version.

File: lib/luajit2/src/vm_ppc.dasc
Comment (cosmo0920) [inline lib/luajit2/src/vm_ppc.dasc:1]: If I'm not wrong, this PR contains only for PowerPC VM code. There is no ppc64le (PowerPC 64bit little endian) support.
ppc64le support should be proceeded (or stucked) here: https://github.com/LuaJIT/LuaJIT/pull/54

**Review by cosmo0920 — CHANGES_REQUESTED** [review]

File: lib/luajit2/Makefile
Comment (cosmo0920) [inline lib/luajit2/Makefile:21]: This is just my curious. Can we use the recent commit of LuaJIT for ppc64le platform?

I don't have real ppc64le environment but it seems that this luajit2 libraries should be luajit 2.1.0-beta3 version.
This seems to be a bit of old than current fbit using version.

File: lib/luajit2/src/vm_ppc.dasc
Comment (cosmo0920) [inline lib/luajit2/src/vm_ppc.dasc:1]: If I'm not wrong, this PR contains only for PowerPC VM code. There is no ppc64le (PowerPC 64bit little endian) support.
ppc64le support should be proceeded (or stucked) here: https://github.com/LuaJIT/LuaJIT/pull/54

**Review by cosmo0920 — COMMENTED** [review]
This is just my first impressions.

**Comment by tarruda** [comment]
> It's up to @edsiper, I personally think that the footprint of the PR without the luajit2 addition would be much more manageable, especially since I think that at the moment we are not using the actual JIT mode in luajit but I could be wrong about that since I'm not familiar enough with the component.

IIRC JIT is enabled by default, but it should be easy to disable for one platform. Unless the user is doing CPU heavy tasks in a lua filter (which is unlikely) then disabling the JIT probably won't make much of a difference in terms of performance.

**Comment by sumitd2** [comment]
Hi @edsiper Please let me know how you would like to proceeed with luajit2.

**Comment by edsiper** [comment]
my only real concern is about the LuaJIT version proposed (because I know nothing about it and I don't think I am the one to vouch for it), the proposed ones come from OpenResty project while we use upstream LuaJIT, which is again continuously maintained:

- https://github.com/luajit/luajit

@tarruda LuaJIT don't enable JIT by default, can you research a bit between both versions and potential breaking changes for users ?

**Comment by tarruda** [comment]
@edsiper The changes are well documented at https://github.com/openresty/luajit2 and are mainly new added APIs and some internal changes related to string hashing. Doesn't seem anything in openresty's version is backwards incompatible with upstream LuaJIT.

In fact, it seems PPC is already supported in upstream, and has been for a long time

image

I don't see anything in OpenResty's LuaJIT related to IBM Power. I might be missing something, but I don't understand why adding this support requires changing the used LuaJIT. Can you elaborate on that @sumitd2 ?

**Comment by tarruda** [comment]
@sumitd2 have you tested building fluent-bit on PPC64LE without filter_lua enabled? If fluent-bit compiles, then you should be able to enable filter_lua by simply disabling the JIT at build time (pass -DLUAJIT_DIABLE_JIT as a compile time option).

**Comment by edsiper** [comment]
you can disable all LuaJIT stuff with `-DFLB_LUAJIT=OFF -DFLB_FILTER_LUA=OFF`

**Comment by sumitd2** [comment]
@tarruda I used 
**cmake -DFLB_WASM=OFF -DFLB_LUAJIT=OFF -DFLB_FILTER_LUA=OFF -DFLB_TESTS_RUNTIME=On -DFLB_TESTS_INTERNAL=On ..**
It compiled fine and the filter_lua test was enabled, though it failed. Have removed luajit2 from the PR. Let me know how i can fix the failing and the action required check.

**Comment by patrick-stephens** [comment]
@sumitd2 have you got an example of someone building for PPC on Github Actions? As you say we can do builds with QEMU/buildx which is what we do now, I'd just like a pointer at any working images as it's not a target I have much experience with recently.

**Comment by sumitd2** [comment]
@tarruda I applied that patch (modified luajit.cmake manually). Results in compile errors:

```
==== Building LuaJIT 2.1.0-beta3 ====
lj_arch.h:327:2: error: #error "No support for PPC64"
 #error "No support for PPC64"
  ^~~~~
lj_arch.h:494:2: error: #error "No support for little-endian PPC32"
 #error "No support for little-endian PPC32"

```

**Comment by tarruda** [comment]
@sumitd2 I assumed disabling JIT would be enough to make LuaJIT compile on any arch supported by the C compiler, not sure what can be done in this case then

**Comment by sumitd2** [comment]
Hi @patrick-stephens Have a look at:

https://github.com/ansible/receptor/blob/devel/.github/workflows/devel_image.yml#L51
https://github.com/ansible/receptor/actions/runs/2972734524/jobs/4761505405

**Comment by sumitd2** [comment]
Hi team, @tarruda 

We would need to close this PR because fb is not building on power anymore because of wasm. This is the first error I got (I fixed this and there were a couple more):

_cc: error: unrecognized command line option '-mindirect-branch-register'
make[2]: *** [src/wasm/CMakeFiles/vmlib-static.dir/build.make:76: src/wasm/CMakeFiles/vmlib-static.dir/__/__/lib/wasm-micro-runtime-fast-jit-06-29-2022/core/shared/platform/linux/platform_init.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:6952: src/wasm/CMakeFiles/vmlib-static.dir/all] Error 2_

Also, git rebase was failing with merge conflicts when I tried to fix the signoff suffixes on the commits.

We will open a new PR upon getting the build and tests pass when we work on it again. Thank you.

**Comment by cosmo0920** [comment]
> cc: error: unrecognized command line option '-mindirect-branch-register'
make[2]: *** [src/wasm/CMakeFiles/vmlib-static.dir/build.make:76: src/wasm/CMakeFiles/vmlib-static.dir///lib/wasm-micro-runtime-fast-jit-06-29-2022/core/shared/platform/linux/platform_init.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:6952: src/wasm/CMakeFiles/vmlib-static.dir/all] Error 2

Ah, that failures are fixed by https://github.com/fluent/fluent-bit/pull/6036.

**Comment by sumitd2** [comment]
@cosmo0920 I fixed it looking at that PR, then had to fix another one, and then there was another one. It would take some fixing.

**Comment by cosmo0920** [comment]
How about using this patch? WAMR does not support PPC64LE platform. (No mention for ppc64le on their cmake files and ours.)

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index d9531dc1d..8a4757daa 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -564,6 +564,10 @@ if(FLB_WASM)
       message(STATUS "This platform does not provide stdatomic.h that is needed for the FLB_WASM feature so disabled.")
       set(FLB_WASM OFF)
     endif ()
+    if(CMAKE_SYSTEM_PROCESSOR MATCHES "^(powerpc|ppc)64le")
+      message(STATUS "This platform does not supported WAMR building so disabled.")
+      set(FLB_WASM OFF)
+    endif ()
   else ()
     enable_language (ASM)
     FLB_DEFINITION(FLB_HAVE_WASM)
```

**Comment by sumitd2** [comment]
@cosmo0920 This patch will of course do. Shall we do the same for luajit as well?

**Comment by cosmo0920** [comment]
Yes. Also needed similar patch for LUAJIT:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index d9531dc1d..9a0224525 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -907,8 +907,13 @@ endif()
 # LuaJIT (Scripting Support)
 # ==========================
 if(FLB_LUAJIT)
-  include(cmake/luajit.cmake)
-  FLB_DEFINITION(FLB_HAVE_LUAJIT)
+  if(CMAKE_SYSTEM_PROCESSOR MATCHES "^(powerpc|ppc)64le")
+    message(STATUS "This platform does not support LuaJIT so disabled.")
+    set(FLB_LUAJIT OFF)
+  else ()
+    include(cmake/luajit.cmake)
+    FLB_DEFINITION(FLB_HAVE_LUAJIT)
+  endif ()
 endif()
 
 # PostgreSQL
```

**Comment by cosmo0920** [comment]
@sumitd2 Could you use that approach to detect ppc64le platform? https://github.com/cosmo0920/fluent-bit/commit/77f53a114b6b9410ffad262eb6eae331b298342b
We should use the specific cmake file for ppc64le handling.

**Comment by sumitd2** [comment]
> @sumitd2 Could you use this approach to detect ppc64le platform? [cosmo0920@77f53a1](https://github.com/cosmo0920/fluent-bit/commit/77f53a114b6b9410ffad262eb6eae331b298342b) We should use the specific cmake file for ppc64le handling to keep maintanability.

@cosmo0920 Done

**Comment by sumitd2** [comment]
Hi Team, please merge this PR. Thanks.

File: src/flb_utils.c
Comment (leonardo-albertovich) [inline src/flb_utils.c:709]: I would like to understand this part before this is merged so I'll mark this review as asking for changes just so no one merges it by mistake.

Could you explain the logic behind this change and why is it for ppc64le support?

**Review by leonardo-albertovich — CHANGES_REQUESTED** [review]

File: src/flb_utils.c
Comment (sumitd2) [reply src/flb_utils.c:709]: The values being received for power were in a different range, and needed the logic from the _else if (c > 0xFFFF) {_ part of the code. There were two failing tests, which passed after this change. Thats the best I know.

Comment (leonardo-albertovich) [reply src/flb_utils.c:709]: Could you please explain why would that happen? at first (when I hadn't read your PR message) I thought it could've been due to the endianness difference but considering that supporting little endian is the point of the PR that doesn't make sense (obviously).

So I think we need to properly understand why those cases were failing, what the difference with x86 or arm is and what the correct fix is (even if in the end it's just disabling the block with proper understanding of the reason why (and a comment)).

**Comment by leonardo-albertovich** [comment]
Do you have instructions to build and test this in an emulator such as qemu or bochs?

**Comment by sumitd2** [comment]
> Do you have instructions to build and test this in an emulator such as qemu or bochs?

Unfortunately no

**Comment by patrick-stephens** [comment]
> Do you have instructions to build and test this in an emulator such as qemu or bochs?

I agree, I think we should have a build included as part of this change: otherwise it is just extra complexity in the code base and we may break things in the future without realising. This was my point here: https://github.com/fluent/fluent-bit/pull/5934#issuecomment-1243681762

It may be that we can just add a new target to one of the existing packaging builds: https://github.com/fluent/fluent-bit/tree/master/packaging/distros
These are container-based QEMU builds so we could add a `ppc` option to verify. How are you currently building and testing it @sumitd2 , i.e. what OS, etc.? Can we build for a PPC variant of any of our existing targets easily?

**Comment by sumitd2** [comment]
Hi @patrick-stephens and team, 

We use Redhat UBI 8.x containers inside a ppc64le VM for our power builds.

We agree with your concern that without a PPC build in the CI, it is just extra complexity in the code base and you may break things in the future without realising.  So can you add a Qemu based ppc build yourself, or would you need our help with it? These are two examples of such builds:
https://github.com/ansible/receptor/blob/devel/.github/workflows/devel_image.yml#L51
https://github.com/ansible/receptor/actions/runs/2972734524/jobs/4761505405

Thanks.

**Comment by sumitd2** [comment]
> I can probably do it with a ubi one but it would be better if we can do it with one of the existing targets I think. The builds are all done from the container definitions in packaging/distros
> 
> I do a ubi build for our enterprise offerings and also for the certified RHCC offering here: https://github.com/calyptia/fluent-bit

Hi @patrick-stephens, do you have plans to look including power support in your CI?

**Comment by patrick-stephens** [comment]
> Hi @patrick-stephens, do you have plans to include power support in your CI?

Possibly if it is simple to do and secure, we would need a solution for ephemeral Github Action runners on custom self-hosted VMs - Github currently recommend against using self-hosted runners for public repos due to the security concerns around PRs. It is something I'm looking at more generally for other reasons too (e.g. the ARM builds are incredibly slow).

My first preference would be a slow but functional QEMU build if we can do it on the hosted Github runners. We can then run this build on demand for overnight or formal releases for example rather than on every commit (same as we do for ARM).

It is easier to do in our enterprise offerings built privately which is what I was going to try first and then check.

**Pull request closed without merging**
